i18n: Migrate from @automattic/react-i18n to @wordpress/react-i18n

yuliyan opened the pull request:
#### Changes proposed in this Pull Request

* Use `@wordpress/react-i18n` in `CalypsoI18nProvider` and Gutenboarding's `LocaleContext` components.
* Replace `@automattic/react-i18n` imports with `@wordpress/react-i18n` in all dependant components.
* Update dependencies.
* Resolve type conflicts in `@automattic/react-i18n` package due to type def changes in `@wordpress/i18n@^3.19.0`.

#### Testing instructions

* Review code changes - `CalypsoI18nProvider`, packages dependencies updates, etc.
* Test Gutenboarding and confirm translations are loaded and displayed as expected.
* Test ETK (Editor Site Launch) and confirm translations are loaded and displayed as expected.
* Test translations in packages:
  * Composite Checkout - can be tested on `calypso.live/checkout/{site}`
  * Onboarding - can be tested on `calypso.live/checkout/{site}`, the action buttons `Skip for now`, `Go back`
  * Language Picker & Search - can be tested on `calypso.live/me/account` (Interface language) or `calypso.live/new/language-modal`
  * Domain picker - can be tested on `calypso.live/new/domains`
  * Plans Grid - can be tested on `calypso.live/new/plans`
  * What's New - can be tested on `calypso.live/post/{site}/new`, Options Menu (three dots button in the top bar) -> What's new
  * Focused Launch - can be tested on `calypso.live/page/{site}/{pageid}`, where `{site}` is a non-launched site

yuliyan committed:
i18n: Migrate from @automattic/react-i18n to @wordpress/react-i18n

yuliyan committed:
Deduplicate packages

sirbrillig reviewed (approved):
Good timing! I was just looking at this because I'm working to publish the `composite-checkout` package and its dependency on the private `@automattic/react-i18n` was a blocker. 

Tested with checkout and everything appears to still be translated!

yuliyan committed:
Reuse variable for `localeSlug` prop

p-jackson reviewed (commented):
Still testing this, but I noticed there's a few TS project configs that can be updated with this change too.

e.g. `domain-picker/package.json` has had the `@automattic/react-i18n` dependency removed, so it should also be removed from `domain-picker/tsconfig.json`. The `references` prop in the TS config is just used by the compiler to determine build order so having it there isn't breaking anything, just want to keep those references lists up to date.

So any `package.json` file that's been changed in this PR will probably also require a change to the corresponding `tsconfig.json` (if that package has any TS that is)

p-jackson commented on client/components/calypso-i18n-provider/index.tsx:
```diff
@@ -2,21 +2,23 @@
 const CalypsoI18nProvider: React.FunctionComponent = ( { children } ) => {
 	const [ localeData, setLocaleData ] = React.useState( i18n.getLocale() );
 	const [ localeSlug, setLocaleSlug ] = React.useState( i18n.getLocaleSlug() );
+	const wpI18n = React.useMemo( () => createI18n( localeData ), [ localeSlug ] );
```
Given we're already calling `wpI18n.setLocaleData()` when the locale changes, do we really need to recreate this object when the slug changes?

Using `useRef()` would be a way of having a single I18n instance for the lifetime of the component.

```suggestion
	const wpI18n = React.useRef( createI18n( localeData ) );
```

ciampo replied:
That's what I was thinking too.

And even if we kept the `useMemo` call, shouldn't the hook dependency be `localeData` (instead of `localeSug`?)

p-jackson commented on client/landing/gutenboarding/components/locale-context/index.tsx:
```diff
@@ -51,11 +51,14 @@ export const ChangeLocaleContextConsumer = ChangeLocaleContext.Consumer;
 	const [ contextLocaleData, setContextLocaleData ] = React.useState< LocaleData | undefined >();
 	const [ localeDataLoaded, setLocaleDataLoaded ] = React.useState( false );
+	const localeSlug = contextLocaleData?.[ '' ]?.localeSlug ?? DEFAULT_LOCALE_SLUG;
+	const wpI18n = React.useMemo( () => createI18n( contextLocaleData ), [ localeSlug ] );
```
Similar situation here, do we just want a single instance of I18n for the lifetime of the component by using `useRef()` instead?

ciampo replied:
Similar comment too from my side — `localeSlug`  is listed as a dependency of `useMemo`, but it should probably be `contextLocaleData` ?

yuliyan replied:
See why I think it needs to be memoized by the localeSlug here - https://github.com/Automattic/wp-calypso/pull/51406#discussion_r604757485.

ciampo replied:
Got it! Would you mind adding a comment in code to explain the reason of the memoization and why `localeSlug` is the dependency instead of `contextLocaleData` ?
Without any context, opening that file and seeing the ESlint warning may prompt other folks to make changes to it

p-jackson replied:
> even though the name of the setLocaleData method suggests that it will set the locale data, what it actually does it to merge the passed locale data with the existing one

Whoa! Good to know, thanks

p-jackson commented on packages/react-i18n/src/index.tsx:
```diff
@@ -21,7 +24,17 @@ import type { addFilter, removeFilter, hasFilter, applyFilters } from '@wordpres
  * Default i18n instance.
  */
-const defaultI18n: I18n = { __, _n, _nx, _x, isRTL, setLocaleData };
+const defaultI18n: I18n = {
```
Given `@wordpress/i18n` now exports the default I18n instance, perhaps we can just use that instead of creating our own?

Maybe it doesn't matter since I suspect this a8c package will probably disappear eventually.

yuliyan replied:
Good point. I've updated the code, even though after this PR `@automattic/react-i18n` won't be used anywhere and it should be safe to remove it from the repo. The only reason for updating the package here is to avoid package build failures.

p-jackson reviewed (approved):
Reviewed and tested ETK and onboarding 👍 
I left some comments about the use of `useMemo()`. I don't think it's a blocker for deploying.

Note, the contents of the "whats new" modal isn't translated correctly, but that's due to a bug in the feature, not this change. I'll create an issue or PR, but it's using the `useLocale()` hook (from `@a8c/i18n-utils`) but the `<LocaleProvider>` hasn't been set up first. One of the nice things about the new `@wp/i8n` features is that we should be able to phase out `<LocaleProvider>` and use standard core features instead, since we can now access the current locale slug using `getLocaleData()`.

jsnajdr commented on client/components/calypso-i18n-provider/index.tsx:
```diff
@@ -2,21 +2,23 @@
 			const nextLocaleData = Object.assign( {}, i18n.getLocale() );
 
 			setWpI18nLocaleData( nextLocaleData );
+			wpI18n.setLocaleData( nextLocaleData );
```
This component could be radically simplified, as it currently stores `localeData` at three places and we need only one.

Creating the `wpI18n` instance should not be needed. The locale data are already in the default `i18n` instance and we are setting them with `setWpI18nLocaleData`.

The `localeData` local state can also be removed. It doesn't seem to be used by anything else bug `createI18n( localeData )`. That create call can be removed and even if it couldn't, we could call `createI18n( i18n.getLocale() )`.

yuliyan replied:
We do need to created new `@wordpress/i18n` instance on locale change, because even though the name of the `setLocaleData` method suggests that it will set the locale data, what it actually does it to [merge the passed locale data with the existing one](https://github.com/WordPress/gutenberg/blob/0eb76dcbdb22ef198c690ade215acb56604edd00/packages/i18n/src/create-i18n.js#L163-L176). This means that if we decide to use the same instance across locale changes, it may end up displaying a translations from previous locales for translations entries that are missing in the current locale data, instead of falling back to showing the original (English) string. This behaviour would be observed in case the user switches from Mag-16 to a non-Mag-16 language, as the latter usually have incomplete translations.

The way `@automattic/react-i18n` currently works is that it would create an i18n instance when the locale data changes, but I believe that the reason why it was implemented like that was because `setLocaleData` was not available back then. Currently, it seems to be redundant, which is why I've decided to memoize it just by the locale slug.

As for the locale data state - you're right, it's no longer needed, so I've removed it.

jsnajdr replied:
In that case, why do we use the default `i18n` instance and call `setLocaleData` on it? That way, the default instance will contain a mixture of merged translations after a few locale switches.

yuliyan replied:
> In that case, why do we use the default `i18n` instance and call `setLocaleData` on it? That way, the default instance will contain a mixture of merged translations after a few locale switches.

That's true, but that's basically how it currently works and I don't think there's much we can do with what the current `@wordpress/i18n` API provides. So I'd rather preserve as is currently on production and not change anything in this PR that would alter its behaviour.

However, we should address that in a separate PR and for that we may need to propose adding a new `resetLocaleData` method to `@wordpress/i18n`, or updating the `setLocaleData`, although that could be a breaking change.

jsnajdr replied:
OK, that makes sense. Eventually we should use only one `i18n` instance -- either the global default one or a custom one recreated for each locale. But not both at the same time. That feels like something that should be cleaned up.

> for that we may need to propose adding a new `resetLocaleData` method

Yes, either that or adding a new "merge or rewrite" boolean argument to `setLocaleData`.

yuliyan replied:
I've created a PR with the proposed `setLocaleData` enhancement here - https://github.com/WordPress/gutenberg/pull/30419.

ciampo commented on client/components/calypso-i18n-provider/index.tsx:
```diff
@@ -2,21 +2,23 @@
 import { LocaleProvider, i18nDefaultLocaleSlug } from '@automattic/i18n-utils';
 
 const CalypsoI18nProvider: React.FunctionComponent = ( { children } ) => {
 	const [ localeData, setLocaleData ] = React.useState( i18n.getLocale() );
```
Not sure if introduced by this PR, but on my local machine I see a TypeScript error about `getLocale()`:

```
Property 'getLocale' does not exist on type 'typeof i18nCalypso'
```


Same for the `i18n.on()` and `i18n.off()` calls later in this file

yuliyan replied:
I don't think it was introduced in this PR. I'll address it in a follow up PR.

p-jackson commented:
> Note, the contents of the "whats new" modal isn't translated correctly, but that's due to a bug in the feature, not this change.

I've created a PR that was initially meant to just fix the What's New modal, but it appears to have another problem with the end point. But this PR should fix cases where the `<LocaleProvider>` is forgotten #51567

ciampo reviewed (commented):
Hey @yuliyan, thank you for working on this refactor!

I left a couple of inline comments/questions, here is the rest of them:

- Not all references of `"@wordpress/i18n"` have been updated to version `"^3.19.0"` — many are still referring to "^3.18.0" (`apps/o2-blocks/package.json`, `apps/wpcom-block-editor/package.json`, `packages/design-picker/package.json`, `packages/page-pattern-modal/package.json`, `packages/social-previews/package.json`)

- As @p-jackson pointed out, we should update the `tsconfig.json` files too (removing `react-i18n` from the `references` array)

- Looks like the [`@automattic/design-picker` package](https://github.com/Automattic/wp-calypso/blob/c9fe2bd56261aff7454a828785c74be9a463d17f/packages/design-picker/package.json) is still using `@automattic/react-i18n`

- Maybe it's a silly question, but — If we're moving away from `@automattic/react-i18n`, why is the `packages/react-i18n/src/index.tsx` being updated, instead of deleting the whole package?

yuliyan committed:
Remove locale data state from Calypso i18n provider

ciampo commented on packages/react-i18n/src/index.tsx:
```diff
@@ -178,7 +191,7 @@ function hasTranslation( localeData: LocaleData, singular: string, context?: str
 	const i18n = localeData ? createI18n( localeData ) : defaultI18n;
 	const boundHasTranslation = localeData
-		? ( singular: string, context?: string ) => hasTranslation( localeData, singular, context )
+		? ( singular: string, context?: string ) => doHasTranslation( localeData, singular, context )
```
Now that `@wordpress/i18n` exports `hasTranslation`, why don't we just use that implementation directly?

yuliyan replied:
With the changes introduced in this PR, the package will no longer be used anywhere and will most likely be removed in an immediate PR after this one gets merged. The changes here are only meant to resolve the build errors, caused by package version upgrades.

yuliyan committed:
Use defaultI18n instance @wordpress/i18n in @automattic/react-i18n

yuliyan committed:
Remove packages/react-i18n reference from tsconfigs

yuliyan committed:
Migrate design-picker to @wordpress/react-i18n

yuliyan commented:
> * Not all references of `"@wordpress/i18n"` have been updated to version `"^3.19.0"` — many are still referring to "^3.18.0" (`apps/o2-blocks/package.json`, `apps/wpcom-block-editor/package.json`, `packages/design-picker/package.json`, `packages/page-pattern-modal/package.json`, `packages/social-previews/package.json`)

I wanted to keep the changes in this PR as little as possible and I've only `"@wordpress/i18n"` for those packages that use also `"@wordpress/react-i18n"` because `3.19.0` could be considered somewhat critical for it. And for the packages, that you've listed above, they'll still be using `3.19.0` in Calypso env, because of the deduped lock file, and if they're being used outside of Calypso, they should be fine with `3.18.0`. However, if we want to upgrade them as well, I think it would be better to do so in a separate PR.

----

> * As @p-jackson pointed out, we should update the `tsconfig.json` files too (removing `react-i18n` from the `references` array)

Thanks for the reminder, I've removed them in https://github.com/Automattic/wp-calypso/pull/51406/commits/551aaf2a09b9fec06033ca2f3a1931a97ae5c591.

----

> * Looks like the [`@automattic/design-picker` package](https://github.com/Automattic/wp-calypso/blob/c9fe2bd56261aff7454a828785c74be9a463d17f/packages/design-picker/package.json) is still using `@automattic/react-i18n`

Looks like the packages has been introduced after I've created the PR, so if it wasn't for this ping, I would've missed it. Thanks :) I've added it in https://github.com/Automattic/wp-calypso/pull/51406/commits/a486b13822454505db449ac48baf39ccf97a82a9.

----

> * Maybe it's a silly question, but — If we're moving away from `@automattic/react-i18n`, why is the `packages/react-i18n/src/index.tsx` being updated, instead of deleting the whole package?

My idea is to remove the package in a follow up PR and the only reason for the update here is to avoid build errors in the `build-packages` process.

yuliyan committed:
Add comments on why is wpI18n instance being memoized

jsnajdr commented on client/components/calypso-i18n-provider/index.tsx:
```diff
@@ -2,22 +2,26 @@
 	React.useEffect( () => {
 		const onChange = () => {
 			// Create a new object to ensure the provider sees the data has changed
 			const nextLocaleData = Object.assign( {}, i18n.getLocale() );
```
Here we do not need the `Object.assign` any more. `nextLocaleData` is no longer stored to state or passed as a prop to `I18nProvider` as it used to in the old version. So we don't need to ensure that it's a different object instance.
```js
const nextLocaleData = i18n.getLocale();
```
should be sufficient.

jsnajdr commented on client/components/calypso-i18n-provider/index.tsx:
```diff
@@ -26,11 +30,11 @@ const CalypsoI18nProvider: React.FunctionComponent = ( { children } ) => {
 			i18n.off( 'change', onChange );
 		};
-	}, [] );
+	}, [ localeSlug ] );
```
We don't need this `localeSlug` dependency. All it does is that it unregisters the `i18n` subscription and adds it again, without changing anything.

yuliyan committed:
Remove redundant Object.assign() call

yuliyan committed:
Remove unnecessary useEffect dependency in CalypsoI18nProvider

dlind1 reviewed (approved):
Reviewed and tested the translations, everything seems to be working except for What's New API calls, which is handled in several diffs and https://github.com/Automattic/wp-calypso/pull/51567, LGTM.

jsnajdr commented on client/components/calypso-i18n-provider/index.tsx:
```diff
@@ -2,22 +2,23 @@
+	// which may lead to residue translations from previous locales to remain causing displaying mixed up translations.
+	const wpI18n = React.useMemo( () => createI18n( i18n.getLocale() ), [ localeSlug ] );
+
 	React.useEffect( () => {
```
The linter will complain about a missing `wpI18n` dependency for this hook. Trying to resolve this puzzle, I arrived at this alternative implementation:
```jsx
const createValue = () => ( {
	slug: i18n.getLocaleSlug(),
	i18n: createI18n( i18n.getLocale() ),
} );

const CalypsoI18nProvider: React.FunctionComponent = ( { children } ) => {
	const [ value, setValue ] = React.useState( createValue );

	React.useEffect( () => {
		const onChange = () => {
			setValue( ( prevValue ) => {
				// always sync the global default i18n instance
				setWpI18nLocaleData( i18n.getLocale() );

				if ( prevValue.slug !== i18n.getLocaleSlug() ) {
					// Create a new @wordpress/i18n instance when the locale changes,
					// as `setLocaleData` doesn't replace the entire locale data, but rather merges it with the existing one,
					// which may lead to residue translations from previous locales to remain causing displaying mixed up translations.
					return createValue();
				}

				// merely merge the new locale data into existing instance and don't change state
				prevValue.i18n.setLocaleData( i18n.getLocale() );
				return prevValue;
			} );
		};

		i18n.on( 'change', onChange );

		return () => {
			i18n.off( 'change', onChange );
		};
	}, [] );

	return (
		<LocaleProvider localeSlug={ value.slug || i18nDefaultLocaleSlug }>
			<I18nProvider i18n={ value.i18n }>{ children }</I18nProvider>
		</LocaleProvider>
	);
};
```
It:
- creates a local `i18n` instance
- recreates it on upstream (`i18n-calypso`) locale slug change
- passes it down in `<I18nProvider>`
- synces data changes from upstream to the local `i18n` instance
- synces data changes to the global default `i18n` instance, too. Keeping the bug where a locale change would merge data from two languages together

jsnajdr reviewed (approved):
I posted some suggestions on how to rewrite the `CalypsoI18nProvider`. Similarly, I believe the Gutenboarding `LocaleContext` component could be simplified a lot, too. But both of them are quite complicated React hook puzzles 🙂 They don't need to prevent this awesome migration from being merged.
